# Import necessary Libraries

In [1]:
import pandas as pd

# Read your CSV File witht he extracted data

In [5]:
df = pd.read_csv('Extracted Data.csv')

In [6]:
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        352755   
2      Apple         2021         365817       94680        351002   
3      Tesla         2023          96773       14974        106618   
4      Tesla         2022          81462       12556         82338   
5      Tesla         2021          53821        5519         62131   
6  Microsoft         2023         211915       72361        411976   
7  Microsoft         2022         198270       72738        364840   
8  Microsoft         2021         168088       61271        333779   

   Total Liabilities Cash Flow from Operating Activities  
0             290437                              110543  
1             302083                              122151  
2             287912                              104038  
3              43009                               13256  
4              36440                               14724  
5              30548                              11497   
6             205753                               87582  
7             198298                               89035  
8             191791                               76740

# Analyzing trends with pandas

## Calculating Year-by-Year growth rates for Total Revenue

In [7]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100

In [8]:
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        352755   
2      Apple         2021         365817       94680        351002   
3      Tesla         2023          96773       14974        106618   
4      Tesla         2022          81462       12556         82338   
5      Tesla         2021          53821        5519         62131   
6  Microsoft         2023         211915       72361        411976   
7  Microsoft         2022         198270       72738        364840   
8  Microsoft         2021         168088       61271        333779   

   Total Liabilities Cash Flow from Operating Activities  Revenue Growth (%)  
0             290437                              110543                 NaN  
1             302083                              122151            2.881146  
2             287912                              104038           -7.230275  
3              43009                               13256                 NaN  
4              36440                               14724          -15.821562  
5              30548                              11497           -33.931158  
6             205753                               87582                 NaN  
7             198298                               89035           -6.438902  
8             191791                               76740          -15.222676

## Replacing NaN values with 0

In [9]:
df.fillna(0, inplace=True)
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        352755   
2      Apple         2021         365817       94680        351002   
3      Tesla         2023          96773       14974        106618   
4      Tesla         2022          81462       12556         82338   
5      Tesla         2021          53821        5519         62131   
6  Microsoft         2023         211915       72361        411976   
7  Microsoft         2022         198270       72738        364840   
8  Microsoft         2021         168088       61271        333779   

   Total Liabilities Cash Flow from Operating Activities  Revenue Growth (%)  
0             290437                              110543            0.000000  
1             302083                              122151            2.881146  
2             287912                              104038           -7.230275  
3              43009                               13256            0.000000  
4              36440                               14724          -15.821562  
5              30548                              11497           -33.931158  
6             205753                               87582            0.000000  
7             198298                               89035           -6.438902  
8             191791                               76740          -15.222676

## Calculating Year-by-Year growth rates for Net Income

In [11]:
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100
df.fillna(0, inplace=True)
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        352755   
2      Apple         2021         365817       94680        351002   
3      Tesla         2023          96773       14974        106618   
4      Tesla         2022          81462       12556         82338   
5      Tesla         2021          53821        5519         62131   
6  Microsoft         2023         211915       72361        411976   
7  Microsoft         2022         198270       72738        364840   
8  Microsoft         2021         168088       61271        333779   

   Total Liabilities Cash Flow from Operating Activities  Revenue Growth (%)  \
0             290437                              110543            0.000000   
1             302083                              122151            2.881146   
2             287912                              104038           -7.230275   
3              43009                               13256            0.000000   
4              36440                               14724          -15.821562   
5              30548                              11497           -33.931158   
6             205753                               87582            0.000000   
7             198298                               89035           -6.438902   
8             191791                               76740          -15.222676   

   Net Income Growth (%)  
0               0.000000  
1               2.894995  
2              -5.133112  
3               0.000000  
4             -16.147990  
5             -56.044919  
6               0.000000  
7               0.520999  
8             -15.764800

## Calculating Year-by-Year growth rates for Total Assets

In [14]:
df['Assets Growth (%)'] = df.groupby(['Company'])['Total Assets'].pct_change() * 100
df.fillna(0, inplace=True)
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        352755   
2      Apple         2021         365817       94680        351002   
3      Tesla         2023          96773       14974        106618   
4      Tesla         2022          81462       12556         82338   
5      Tesla         2021          53821        5519         62131   
6  Microsoft         2023         211915       72361        411976   
7  Microsoft         2022         198270       72738        364840   
8  Microsoft         2021         168088       61271        333779   

   Total Liabilities Cash Flow from Operating Activities  Revenue Growth (%)  \
0             290437                              110543            0.000000   
1             302083                              122151            2.881146   
2             287912                              104038           -7.230275   
3              43009                               13256            0.000000   
4              36440                               14724          -15.821562   
5              30548                              11497           -33.931158   
6             205753                               87582            0.000000   
7             198298                               89035           -6.438902   
8             191791                               76740          -15.222676   

   Net Income Growth (%)  Assets Growth (%)  
0               0.000000           0.000000  
1               2.894995           0.048783  
2              -5.133112          -0.496945  
3               0.000000           0.000000  
4             -16.147990         -22.772890  
5             -56.044919         -24.541524  
6               0.000000           0.000000  
7               0.520999         -11.441443  
8             -15.764800          -8.513595

## Calculating Year-by-Year growth rates for Total Liabilities

In [15]:
df['Liabilities Growth (%)'] = df.groupby(['Company'])['Total Liabilities'].pct_change() * 100
df.fillna(0, inplace=True)
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        352755   
2      Apple         2021         365817       94680        351002   
3      Tesla         2023          96773       14974        106618   
4      Tesla         2022          81462       12556         82338   
5      Tesla         2021          53821        5519         62131   
6  Microsoft         2023         211915       72361        411976   
7  Microsoft         2022         198270       72738        364840   
8  Microsoft         2021         168088       61271        333779   

   Total Liabilities Cash Flow from Operating Activities  Revenue Growth (%)  \
0             290437                              110543            0.000000   
1             302083                              122151            2.881146   
2             287912                              104038           -7.230275   
3              43009                               13256            0.000000   
4              36440                               14724          -15.821562   
5              30548                              11497           -33.931158   
6             205753                               87582            0.000000   
7             198298                               89035           -6.438902   
8             191791                               76740          -15.222676   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  
0               0.000000           0.000000                0.000000  
1               2.894995           0.048783                4.009820  
2              -5.133112          -0.496945               -4.691095  
3               0.000000           0.000000                0.000000  
4             -16.147990         -22.772890              -15.273547  
5             -56.044919         -24.541524              -16.169045  
6               0.000000           0.000000                0.000000  
7               0.520999         -11.441443               -3.623276  
8             -15.764800          -8.513595               -3.281425

In [19]:
df['Cash Flow from Operating Activities'] = pd.to_numeric(df['Cash Flow from Operating Activities'], errors='coerce')


## Calculating Year-by-Year growth rates for Total Cash Flow Growth From Operations

In [20]:
df['Cash Flow Growth(%) from operations'] = df.groupby(['Company'])['Cash Flow from Operating Activities'].pct_change() * 100
df.fillna(0, inplace=True)
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        352755   
2      Apple         2021         365817       94680        351002   
3      Tesla         2023          96773       14974        106618   
4      Tesla         2022          81462       12556         82338   
5      Tesla         2021          53821        5519         62131   
6  Microsoft         2023         211915       72361        411976   
7  Microsoft         2022         198270       72738        364840   
8  Microsoft         2021         168088       61271        333779   

   Total Liabilities  Cash Flow from Operating Activities  Revenue Growth (%)  \
0             290437                             110543.0            0.000000   
1             302083                             122151.0            2.881146   
2             287912                             104038.0           -7.230275   
3              43009                              13256.0            0.000000   
4              36440                              14724.0          -15.821562   
5              30548                                  0.0          -33.931158   
6             205753                              87582.0            0.000000   
7             198298                              89035.0           -6.438902   
8             191791                              76740.0          -15.222676   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
0               0.000000           0.000000                0.000000   
1               2.894995           0.048783                4.009820   
2              -5.133112          -0.496945               -4.691095   
3               0.000000           0.000000                0.000000   
4             -16.147990         -22.772890              -15.273547   
5             -56.044919         -24.541524              -16.169045   
6               0.000000           0.000000                0.000000   
7               0.520999         -11.441443               -3.623276   
8             -15.764800          -8.513595               -3.281425   

   Cash Flow Growth(%) from operations  
0                             0.000000  
1                            10.500891  
2                           -14.828368  
3                             0.000000  
4                            11.074231  
5                             0.000000  
6                             0.000000  
7                             1.659017  
8                           -13.809176

# Save the file

In [29]:
df.to_csv('Financial Data.csv', index=False)

# Summarize your Data based on company-wise metrics

In [24]:
summary_company_wise = df.groupby('Company').agg({
    'Total Revenue': 'mean',  # Average revenue over the years
    'Net Income': 'mean',     # Average net income over the years
    'Total Assets': 'mean',   # Average total assets
    'Total Liabilities': 'mean',  # Average total liabilities
    'Cash Flow from Operating Activities': 'mean',  # Average cash flow
    'Revenue Growth (%)': 'mean',  # Average revenue growth
    'Net Income Growth (%)': 'mean',  # Average net income growth
    'Assets Growth (%)': 'mean',  # Average assets growth
    'Liabilities Growth (%)': 'mean',  # Average liabilities growth
    'Cash Flow Growth(%) from operations': 'mean'  # Average cash flow growth
})

In [25]:
summary_company_wise

Total Revenue    Net Income   Total Assets  Total Liabilities  \
Company                                                                    
Apple      381143.333333  97159.333333  352113.333333      293477.333333   
Microsoft  192757.666667  68790.000000  370198.333333      198614.000000   
Tesla       77352.000000  11016.333333   83695.666667       36665.666667   

           Cash Flow from Operating Activities  Revenue Growth (%)  \
Company                                                              
Apple                            112244.000000           -1.449710   
Microsoft                         84452.333333           -7.220526   
Tesla                              9326.666667          -16.584240   

           Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
Company                                                                       
Apple                  -0.746039          -0.149388               -0.227092   
Microsoft              -5.081267          -6.651679               -2.301567   
Tesla                 -24.064303         -15.771471              -10.480864   

           Cash Flow Growth(%) from operations  
Company                                         
Apple                                -1.442492  
Microsoft                            -4.050053  
Tesla                                 3.691410

# Summarize your Data based on year-wise metrics

In [27]:
summary_year_wise = df.groupby('Fiscal Year').agg({
    'Total Revenue': 'mean',  # Average revenue over the years
    'Net Income': 'mean',     # Average net income over the years
    'Total Assets': 'mean',   # Average total assets
    'Total Liabilities': 'mean',  # Average total liabilities
    'Cash Flow from Operating Activities': 'mean',  # Average cash flow
    'Revenue Growth (%)': 'mean',  # Average revenue growth
    'Net Income Growth (%)': 'mean',  # Average net income growth
    'Assets Growth (%)': 'mean',  # Average assets growth
    'Liabilities Growth (%)': 'mean',  # Average liabilities growth
    'Cash Flow Growth(%) from operations': 'mean'  # Average cash flow growth
})

In [28]:
summary_year_wise

Total Revenue    Net Income   Total Assets  Total Liabilities  \
Fiscal Year                                                                  
2021         195908.666667  53823.333333  248970.666667      170083.666667   
2022         224686.666667  61699.000000  266644.333333      178940.333333   
2023         230657.666667  61443.333333  290392.333333      179733.000000   

             Cash Flow from Operating Activities  Revenue Growth (%)  \
Fiscal Year                                                            
2021                                60259.333333          -18.794703   
2022                                75303.333333           -6.459773   
2023                                70460.333333            0.000000   

             Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
Fiscal Year                                                                     
2021                    -25.647610         -11.184021               -8.047188   
2022                     -4.243999         -11.388517               -4.962335   
2023                      0.000000           0.000000                0.000000   

             Cash Flow Growth(%) from operations  
Fiscal Year                                       
2021                                   -9.545848  
2022                                    7.744713  
2023                                    0.000000

# Key Insights

#### The data provides a summary of Total Revenue, Net Income, Total Assets, Total Liabilities, and Cash Flow from Operating Activities for three companies (Apple, Microsoft, Tesla) as well as year-over-year changes for various fiscal years (2021, 2022, and 2023)

- Apple's performance has been relatively stable, but it has seen a small decline in key metrics over the period. Revenue, net income, assets, and cash flow from operations have all seen marginal declines, indicating that while the company remains strong, its growth has slightly contracted.
- Microsoft has experienced a significant decline across all growth metrics, with revenue and net income shrinking the most. While its financial position is still strong overall, the decline in assets and cash flow from operations indicates a period of contraction.
- Tesla’s revenue, net income, and assets have declined sharply over the period, reflecting substantial contraction. However, despite these declines, Tesla has managed to grow its cash flow from operating activities, which is a positive signal for its operations efficiency.
- 2021 was a challenging year, with significant declines in revenue, net income, and other financial metrics. It reflects an overall contraction for companies in this dataset during that year.
- 2022 saw some stabilization compared to 2021, though growth rates were still negative for most metrics. Notably, cash flow from operating activities grew by 7.74%, indicating improved operational efficiency across companies.
- 2023 shows no significant changes in growth metrics, suggesting a period of stability after the fluctuations of previous years. Companies seem to have reached a point of equilibrium, with no growth or decline in revenue, income, assets, and liabilities.

### Overall 
- Apple and Microsoft show stability but slight declines across most financial metrics, indicating mature markets with slower growth.
- Tesla experienced the sharpest declines in revenue, net income, and assets, reflecting either market volatility or company-specific challenges. However, its positive growth in cash flow from operations is a positive sign of operational health.
- 2021 was the most challenging year across all companies, with steep declines in revenue and profitability.
- 2022 showed some recovery, particularly in cash flow growth.
- 2023 appears to be a year of stability with little to no change in the financial metrics.